### ISSUE TITLE
Remove/uninstall Proctoring Extension

### ISSUE THREAD

**Issue by holtmeier** [oat-sa/extension-tao-proctoring#350]
Hi,
is there a method to completely remove the proctoring extension? I've deleted the two lines for testcenter and proctoring in composer.json, but ended up with errors. It would be great to have some advice on how to uninstall correctly.
Thanx in advance!
Stephan

**Comment by krampstudio** [comment]
Hi,

Removing the extension using composer will remove the source code, but the extension needs to be _unregistered_ from tao.
1. Restore the source code (using composer)
2. Install the  extension [taoDevTools](https://github.com/oat-sa/extension-tao-devtools).
3. In the "extension manager" you'll have new buttons to uninstall extensions, see 
![TAO Extension Manager with dev tools](http://pix.toile-libre.org/upload/original/1493123920.png)
4. Remove the source code using composer (optional, you can have uninstalled extensions)

Uninstalling extension is still experimental and I discourage you to do it on a production env.

**Comment by holtmeier** [comment]
Thank you! Ok, I did this. First I uninstalled testcenter. After that, I saw the uninstall-button in the row of the proctoring extension. I finally removed both source codes with "composer remove". But now, I'm not able to login. There's a clean white screen and in browser console I get the message "Failed to load resource: the server responded with a status of 500 (Internal Server Error)". I'm not sure, where I can find a more specific error description.

**Comment by krampstudio** [comment]
You should get the error in the PHP logs or in the TAO logs.

To configure the tao logs, pleas edit the file `config/generis/log.conf.php` and configure a file appender : 
```php
return [
    array(
        'class' => 'SingleFileAppender',
        'threshold' => 4,
        'max_file_size' => 1048576, // 1Mb
        'rotation-ratio' => .5,
        'file' => dirname(__FILE__) . '/../../log/tao.log',
        'format' => '[%t] %m',
        'prefix' => ''
    ]
];
```
Do not forget to adapt the path to your system and to add the correct permissions to the folder so the web app can write into it. Then you'll see the error.

May be @jbout knows this issue.

**Comment by holtmeier** [comment]
Hmm, I still have trouble.

After removing proctoring-extension, I can't login as a testtaker. I get a 404 error-message:

![error](https://cloud.githubusercontent.com/assets/1990346/25406018/91d77630-2a05-11e7-80df-5b59feb52d7b.jpg)

The tao log tells me this:

[1493151497] Successful login of user 'test123'.
[1493151497] Invoking tao_actions_Main::entry by http://*******rdf#i14909*********
[1493151497] **Access denied** to oat\taoCe\actions\Main@index for user 'http://********.rdf#i14909******'
[1493151497] Extension or controller not found

When I re-install proctoring-extension, I can login again. But I can't start any test, because I have to authorize it first. And that's exactly,  what I don't want.

**Comment by krampstudio** [comment]
To disable the proctoring authorization page, you have to replace the configuration file `config/taoDelivery/authorization.conf.php` by `taoDelivery/config/default/authorization.conf.php`

**Comment by jbout** [comment]
Hello,

Indeed uninstalling extensions is still an experimental feature and the extensions taoProctoring is still missing multiple post uninstall scripts. Unfortunately the operation krampstudio mentioned is only one of around 10 steps required to make sure Tao behaves exactly as did before the installation of taoProctoring.

If this is an experimental server, I would suggest to export your content and simply reinstall the server.

If that is not possible I would need to know the exact versions of taoProctoring and taoTestCenters that you has installed, in order to step you through the actions required.

Joel

**Comment by holtmeier** [comment]
Hi jbout,

unfortunately it's not an experimental server. It would be much better for me to step through the actions required to do a clean remove.

I'm on the latest development release (I know: not optimal for a production system, but I needed some features...):

taoResultServer -> 2.12.1

taoTestCenters -> It's not installed anymore. I can reinstall the current version.

In a hopeful way...
Stephan

**Comment by jbout** [comment]
Do you know the exact date or the last commit of the  development branch you used for taoProctoring?

There have been a few modifications lately

**Comment by holtmeier** [comment]
I could simply do a composer update?
It's sprint50 1/2.

**Comment by jbout** [comment]
Since you used the develop branch to install taoProctoring, we don't know which version you had installed exactly, so we need to go by date:

When did you run the composer update that added taoProctoring?

**Comment by holtmeier** [comment]
Oh, I'm reading my own post above. I don't know, why I send you the version of taoResultServer! My foult. I have installed tao Proctoring 4.18.0. I believe, the latest commit is from 5 days ago.

**Comment by holtmeier** [comment]
I come from the 3.x-Branch and switched to 4.x accidentally.

**Comment by holtmeier** [comment]
Hi jbout. Would it be "safe" to follow kramstudios instruction (overwriting config/taoDelivery/authorization.conf.php) to get access to the tests again and follow your further steps later? Or would you recommend me to wait?
I don't want to be pushy. I'm very happy for your support. I just need a working system until tomorrow.

**Comment by jbout** [comment]
As it is easier for me to write the required uninstall scripts, than to walk you through them, I did just that:

https://github.com/oat-sa/extension-tao-proctoring/tree/feature/proctoring-uninstall

In order to get your system working again, switch to this branch, reinstall taoProctoring and then uninstall it again. These scripts are untested and could destroy your system, so I recommend a backup (as I would before every major modification of your system, including installing new extensions).

Please let me know if they worked for you

**Comment by holtmeier** [comment]
Oh wow! I'll do that. I let you know what happens :-) thanks!

**Comment by holtmeier** [comment]
Well, I'm struggling with the easiest things. How do I switch to the "feature/proctoring-uninstall"-branch? I tried this in composer.json, but probably I use the wrong syntax.

**Comment by holtmeier** [comment]
Solved it.

This is the correct line:
"oat-sa/extension-tao-proctoring": "dev-feature/proctoring-uninstall"

**Comment by holtmeier** [comment]
Success!

I uninstalled taoProctoring and don't experience any problems. I can access my test again and everything else seams to work as expected.

Again: Thank you very much for your help and for your engagement!

**Comment by jbout** [comment]
You might also want to uninstall taoDevTools as it is not intended for production environments.

**Comment by holtmeier** [comment]
Yes, I'll do this. After removing the Source Code of proctoring, an error occurred:

[1493270357] Access to tao::tao_actions_Main::index denied for user ''
[1493270357] php error(1) in /var/www/html/***********/package-tao/config/tao/entrypoint.conf.php@14: Class 'oat\taoProctoring\model\entrypoint\ProctoringEntryPoint' not found

Therefore, I had to reinstall Proctoring extension (just the Code). Maybe this helps you to optimize the uninstall script.

**Comment by holtmeier** [comment]
Hi @jbout, hi @krampstudio,

after doing the uninstall, I experience a new error, while importing a list of testtakers. The problem seems to be, that the import-function expects the uploaded csv-file in a different location. The upload itself works as expected. The file is located unter /tmp in my tao-folder. But when the function tries to open this csv-file it searches in the wrong place. Looks like a little bit like windows-path.

![err500](https://cloud.githubusercontent.com/assets/1990346/25518644/e7d9d5e6-2bf4-11e7-8d7c-ee83aac9d695.jpg)

Is it possible, that the uninstall procedure has changed a PATH-Variable, which causes this problem now? Normally, I would say no, but I haven't done anything else.

**Comment by jbout** [comment]
* Entrypoints can currently only be disabled not removed, this will require additional work.
* This error seems unrelated to the proctoring extension and I would suggest opening a separate ticket for it.

**Comment by boajer** [comment]
@holtmeier Import related error has been fixed in `tao-core` extension `v9.2.1`
